### PR TITLE
OpenStack Firewall as a Service

### PR DESCRIPTION
**Pull request by ggiamarchi**
**Prerequisites**
- [x] Need PR https://github.com/rackspace/gophercloud/pull/342 to be merged first

**To-do remaining**
- [x] Refactor a bit for code style consistency
- [x] Fix corner case bugs that fail some acceptance tests

### CODE REVIEW

**Comment by ggiamarchi** [comment]
Work is done here

**Comment by jrperritt** [comment]
Cool. I'll try to find some time tomorrow to review.

File: builtin/providers/openstack/resource_openstack_fw_firewall_v2.go [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+
+	"github.com/hashicorp/terraform/helper/resource"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/rackspace/gophercloud"
+	"github.com/rackspace/gophercloud/openstack/networking/v2/extensions/fwaas/firewalls"
+)
+
+func resourceFWFirewallV2() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceFirewallCreate,
```
Comment (jrperritt) [inline builtin/providers/openstack/resource_openstack_fw_firewall_v2.go]: We'll want to add versions to these function names like in the other resource files.

File: builtin/providers/openstack/resource_openstack_fw_firewall_v2.go [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+	log.Printf("[DEBUG] Retrieve information about firewall: %s", d.Id())
+
+	config := meta.(*Config)
+	networkingClient, err := config.networkingV2Client(d.Get("region").(string))
+	if err != nil {
+		return fmt.Errorf("Error creating OpenStack networking client: %s", err)
+	}
+
+	firewall, err := firewalls.Get(networkingClient, d.Id()).Extract()
+	if err != nil {
```
Comment (jrperritt) [inline builtin/providers/openstack/resource_openstack_fw_firewall_v2.go]: Can we use `CheckDeleted` here? It's in `util.go`.

File: builtin/providers/openstack/resource_openstack_fw_firewall_v2.go [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+		Target:     "ACTIVE",
+		Refresh:    WaitForFirewallActive(networkingClient, d.Id()),
+		Timeout:    30 * time.Second,
+		Delay:      0,
+		MinTimeout: 2 * time.Second,
+	}
+
+	_, err = stateConf.WaitForState()
+
+	return firewalls.Update(networkingClient, d.Id(), opts).Err
```
Comment (jrperritt) [inline builtin/providers/openstack/resource_openstack_fw_firewall_v2.go]: This line should be before `stateConf` is defined. And up there it should be:

```
err  = firewalls.Update(networkingClient, d.Id(), opts).Err
if err != nil {
    // ....
}
```

Then, the return statement should be something like:
`return resourceFirewallV2Read()`

File: builtin/providers/openstack/resource_openstack_fw_firewall_v2.go [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+
+	if d.HasChange("admin_state_up") {
+		adminStateUp := d.Get("admin_state_up").(bool)
+		opts.AdminStateUp = &adminStateUp
+	}
+
+	log.Printf("[DEBUG] Updating firewall with id %s: %#v", d.Id(), opts)
+
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{"PENDING_CREATE", "PENDING_UPDATE"},
```
Comment (jrperritt) [inline builtin/providers/openstack/resource_openstack_fw_firewall_v2.go]: Is `PENDING_CREATE` a valid state when updating? Or is this for when a user makes an update request while the firewall is still being created?

File: builtin/providers/openstack/resource_openstack_fw_firewall_v2.go (hunk 3 above)
Comment (jrperritt) [inline builtin/providers/openstack/resource_openstack_fw_firewall_v2.go]: Oh, I see. You're waiting for a pending create or update request to finish before calling this update request. Then perhaps just move the `Update` function one lineup and add the return statement above.

File: builtin/providers/openstack/resource_openstack_fw_policy_v2.go [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+					return hashcode.String(v.(string))
+				},
+			},
+		},
+	}
+}
+
+func resourceFirewallPolicyCreate(d *schema.ResourceData, meta interface{}) error {
+
+	// TODO To remove
```
Comment (jrperritt) [inline builtin/providers/openstack/resource_openstack_fw_policy_v2.go]: Should this still be here?

File: builtin/providers/openstack/resource_openstack_fw_rule_v2.go [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+
+	config := meta.(*Config)
+	networkingClient, err := config.networkingV2Client(d.Get("region").(string))
+	if err != nil {
+		return fmt.Errorf("Error creating OpenStack networking client: %s", err)
+	}
+
+	rule, err := rules.Get(networkingClient, d.Id()).Extract()
+
+	if err != nil {
```
Comment (jrperritt) [inline builtin/providers/openstack/resource_openstack_fw_rule_v2.go]: `CheckDeleted`

**Comment by jrperritt** [comment]
Just a few comments from me. I can't run the tests since I don't have FWaaS, but hopefully when this gets merged others will use it and report any issues. Look good to me though.

File: builtin/providers/openstack/resource_openstack_fw_firewall_v2.go (hunk 1 above)
Comment (ggiamarchi) [inline builtin/providers/openstack/resource_openstack_fw_firewall_v2.go]: Done bd55765d7a3364cfcc3b04311cee119a96363419.

File: builtin/providers/openstack/resource_openstack_fw_firewall_v2.go (hunk 2 above)
Comment (ggiamarchi) [inline builtin/providers/openstack/resource_openstack_fw_firewall_v2.go]: Done 395016b78f36800d320f5ad81f73a9ce6d6785b2

File: builtin/providers/openstack/resource_openstack_fw_firewall_v2.go (hunk 3 above)
Comment (ggiamarchi) [inline builtin/providers/openstack/resource_openstack_fw_firewall_v2.go]: Done 3321c74886ef75541e9ca5ee200b220bb639b716

File: builtin/providers/openstack/resource_openstack_fw_firewall_v2.go (hunk 4 above)
Comment (ggiamarchi) [inline builtin/providers/openstack/resource_openstack_fw_firewall_v2.go]: This is exactly for when a user makes an update request while the firewall is still being created.

File: builtin/providers/openstack/resource_openstack_fw_policy_v2.go (hunk 5 above)
Comment (ggiamarchi) [inline builtin/providers/openstack/resource_openstack_fw_policy_v2.go]: Oups ! It was just for debugging. Deleted 2a5ff8d5da2ee2bb0f169bb0e13a1283d8bbffcb

**Comment by ggiamarchi** [comment]
@julienvey Can you run acceptance tests and validate the behaviour ?

**Comment by ggiamarchi** [comment]
@jrperritt Fixes about your comments are done.

**Comment by julienvey** [comment]
Acceptance tests are sucessful 

Here is the full log http://paste.openstack.org/show/178228/

**Comment by jrperritt** [comment]
Thank you for running and confirming the tests, @julienvey .
LGTM :rocket:
